[CPP20] no class template named 'rebind' in 'class std::allocator<Trajectory*>'

Log: https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc12/CMSSW_13_2_CPP20_X_2023-06-26-1100/TrackingTools/TrajectoryCleaning

```
>> Compiling  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/7fdcdd9715abab468bce186ba18d5404/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_13_2_CPP20_X_2023-06-26-1100/src/TrackingTools/TrajectoryCleaning/src/TrajectoryCleanerFactory.cc
/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DEIGEN_DONT_PARALLELIZE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_13_2_CPP20_X_2023-06-26-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_13_2_CPP20_X_2023-06-26-1100' -I/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/7fdcdd9715abab468bce186ba18d5404/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_13_2_CPP20_X_2023-06-26-1100/src -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/pcre/8.43-37eb2e8b73bab83d6645ecfd5d73dcaa/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/boost/1.80.0-757a7857f4613206a3c6b86989ccf231/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/clhep/2.4.6.0-8b07d134e1b552a3656874209f04981b/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/gsl/2.6-f316a321a173f181b66df52be79d894b/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/hls/2019.08-b8a1533230929513077d97b1507ff465/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/lcg/root/6.28.05-355830244c8fcf8d90c6a07d650d5896/include -isystem/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/tbb/v2021.9.0-903f4083af3d240bceee9925736be659/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/cms/vdt/0.4.3-5a80085534117eaccb28e669c6da4b6f/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/zlib/1.2.11-51072030b7f93c3ac6c4235f21e413cb/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/eigen/82dd3710dac619448f50331c1d6a35da673f764a-b4fe5ef46e60c6c470cd12fab50ff23b/include/eigen3 -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/fmt/8.0.1-54e94b39f5cf29341bb9c4765764e1ca/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/OpenBLAS/0.3.15-15504a5c800a9ecbbc2befbb991bead5/include -I/data/cmsbld/jenkins/workspace/build-any-ib/w/el8_amd64_gcc12/external/tinyxml2/6.2.0-d17873b4d6a42a43226cf689f82ec1ef/include -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -msse3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-deprecated-copy -Wno-unused-parameter -Wunused -Wparentheses -Wno-deprecated -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr  -fPIC  -MMD -MF tmp/el8_amd64_gcc12/src/TrackingTools/TrajectoryCleaning/src/TrackingToolsTrajectoryCleaning/TrajectoryCleanerFactory.cc.d /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/7fdcdd9715abab468bce186ba18d5404/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_13_2_CPP20_X_2023-06-26-1100/src/TrackingTools/TrajectoryCleaning/src/TrajectoryCleanerFactory.cc -o tmp/el8_amd64_gcc12/src/TrackingTools/TrajectoryCleaning/src/TrackingToolsTrajectoryCleaning/TrajectoryCleanerFactory.cc.o
In file included from /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/7fdcdd9715abab468bce186ba18d5404/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_13_2_CPP20_X_2023-06-26-1100/src/TrackingTools/TrajectoryCleaning/src/TrajectoryCleanerBySharedHits.cc:5:
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/7fdcdd9715abab468bce186ba18d5404/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_13_2_CPP20_X_2023-06-26-1100/src/TrackingTools/TrajectoryCleaning/src/OtherHashMaps.h: In instantiation of 'class cmsutil::SimpleAllocHashMultiMap<const TrackingRecHit*, Trajectory*, {anonymous}::HashByDetId, {anonymous}::EqualsBySharesInput>':
/data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/7fdcdd9715abab468bce186ba18d5404/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_13_2_CPP20_X_2023-06-26-1100/src/TrackingTools/TrajectoryCleaning/src/TrajectoryCleanerBySharedHits.cc:33:15:   required from here
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/7fdcdd9715abab468bce186ba18d5404/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_13_2_CPP20_X_2023-06-26-1100/src/TrackingTools/TrajectoryCleaning/src/OtherHashMaps.h:116:61: error: no class template named 'rebind' in 'class std::allocator<Trajectory*>'
   116 |     typedef typename Alloc::template rebind<KeyItem>::other KeyItemAllocator;
      |                                                             ^~~~~~~~~~~~~~~~
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/7fdcdd9715abab468bce186ba18d5404/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_13_2_CPP20_X_2023-06-26-1100/src/TrackingTools/TrajectoryCleaning/src/OtherHashMaps.h:117:63: error: no class template named 'rebind' in 'class std::allocator<Trajectory*>'
   117 |     typedef typename Alloc::template rebind<KeyItem *>::other KeyItemPtrAllocator;
      |                                                               ^~~~~~~~~~~~~~~~~~~
  /data/cmsbld/jenkins/workspace/build-any-ib/w/tmp/BUILDROOT/7fdcdd9715abab468bce186ba18d5404/opt/cmssw/el8_amd64_gcc12/cms/cmssw/CMSSW_13_2_CPP20_X_2023-06-26-1100/src/TrackingTools/TrajectoryCleaning/src/OtherHashMaps.h:118:63: error: no class template named 'rebind' in 'class std::allocator<Trajectory*>'
   118 |     typedef typename Alloc::template rebind<ValueItem>::other ValueItemAllocator;
      |                                                               ^~~~~~~~~~~~~~~~~~
  gmake: *** [tmp/el8_amd64_gcc12/src/TrackingTools/TrajectoryCleaning/src/TrackingToolsTrajectoryCleaning/TrajectoryCleanerBySharedHits.cc.o] Error 1
```

`std::allocator::rebind` was deprecated in C++17, removed in C++20.

--- a/TrackingTools/TrajectoryCleaning/src/OtherHashMaps.h
+++ b/TrackingTools/TrajectoryCleaning/src/OtherHashMaps.h
@@ -113,9 +113,10 @@ namespace cmsutil {
     value_iterator values(K const &key);
 
   private:
-    typedef typename Alloc::template rebind<KeyItem>::other KeyItemAllocator;
-    typedef typename Alloc::template rebind<KeyItem *>::other KeyItemPtrAllocator;
-    typedef typename Alloc::template rebind<ValueItem>::other ValueItemAllocator;
+    using AllocTraits = std::allocator_traits<Alloc>;
+    using KeyItemAllocator = typename AllocTraits::template rebind_alloc<KeyItem>;
+    using KeyItemPtrAllocator = typename AllocTraits::template rebind_alloc<KeyItem *>;
+    using ValueItemAllocator = typename AllocTraits::template rebind_alloc<ValueItem>;
 
     // --- buckets ---
     size_t bucketSize_, bucketCapacity_;
@@ -256,7 +257,7 @@ namespace cmsutil {
       nextKeyItem_ = *currentKeyRow_;
       keyEndMarker_ = nextKeyItem_ + keyRowSize_;
     }
-    keyAlloc_.construct(nextKeyItem_, KeyItem(next, key, nullptr));
+    std::allocator_traits<KeyItemAllocator>::construct(keyAlloc_, nextKeyItem_, KeyItem(next, key, nullptr));
     nextKeyItem_++;
     return (nextKeyItem_ - 1);
   }
@@ -273,7 +274,7 @@ namespace cmsutil {
       nextValueItem_ = *currentValueRow_;
       valueEndMarker_ = nextValueItem_ + valueRowSize_;
     }
-    valueAlloc_.construct(nextValueItem_, ValueItem(next, value));
+    std::allocator_traits<ValueItemAllocator>::construct(valueAlloc_, nextValueItem_, ValueItem(next, value));
     nextValueItem_++;
     return (nextValueItem_ - 1);
   }

